fix: add rolling window override for slots

hackice20 opened the pull request:
## What does this PR do?

- Adds `disableRollingWindowAdjustment` to the v2 Slots API (2024-09-04) so callers can return slots strictly within the requested start/end range instead of the 1-month backward adjustment for rolling-window event types.
- Wires the new flag through public input DTO, controller docs (ApiQuery), TRPC schema/types, and slot computation.
- Keeps existing rolling window logic unchanged unless the flag is explicitly set.
- Also retains prior embed security fix (fail-closed when managed embed client ID missing).


- Fixes  #25405

## Visual Demo (For contributors especially)

- N/A API change

## OUTPUT 

Without `disableRollingWindowAdjustment`:
```
curl -s -H "cal-api-version: 2024-09-04" \
  "http://localhost:3002/v2/slots?eventTypeId=2&start=2025-12-09&end=2025-12-11&timeZone=UTC&format=range"
```
Result:
```
Status: success
Dates: 2025-12-09, 2025-12-10, 2025-12-11
Total slots: 48
```

With `disableRollingWindowAdjustment=true`:
```
curl -s -H "cal-api-version: 2024-09-04" \
  "http://localhost:3002/v2/slots?eventTypeId=2&start=2025-12-09&end=2025-12-11&timeZone=UTC&disableRollingWindowAdjustment=true&format=range"
```
Result:
```
Status: success
Data keys: ['2025-12-09', '2025-12-10', '2025-12-11']
First date slots: 16
Total slots: 48
```
Partial JSON sample (with the flag):
```
{
  "data": {
    "2025-12-09": [
      { "start": "2025-12-09T09:00:00.000Z", "end": "2025-12-09T09:30:00.000Z" },
      { "start": "2025-12-09T09:30:00.000Z", "end": "2025-12-09T10:00:00.000Z" },
      { "start": "2025-12-09T10:00:00.000Z", "end": "2025-12-09T10:30:00.000Z" },
      ...
    ],
    "2025-12-10": [...],
    "2025-12-11": [...]
  }
}
```
## Mandatory Tasks (DO NOT REMOVE)

- [x ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
- Env: Ensure API v2 runs with standard env (DATABASE_*, REDIS_URL, JWT_SECRET, NEXTAUTH_SECRET, etc.).
- Minimal data: An existing event type; example used: `eventTypeId=2`.
- Requests:
  - Without flag (old behavior):
    ```
    curl -s -H "cal-api-version: 2024-09-04" \
      "http://localhost:3002/v2/slots?eventTypeId=2&start=2025-12-09&end=2025-12-11&timeZone=UTC&format=range"
    ```
  - With flag (new behavior):
    ```
    curl -s -H "cal-api-version: 2024-09-04" \
      "http://localhost:3002/v2/slots?eventTypeId=2&start=2025-12-09&end=2025-12-11&timeZone=UTC&disableRollingWindowAdjustment=true&format=range"
    ```
- Expected: Both returned (for this event type without rolling window) `Dates: 2025-12-09, 2025-12-10, 2025-12-11; Total slots: 48`. For event types with rolling window, the new flag should prevent the 1-month backward adjustment and keep results strictly within the requested range.

## Checklist

- I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code does follow the style guidelines of this project
- I have commented my code, particularly in hard-to-understand areas
- I have checked if my changes generate no new warnings

hackice20 committed:
Add rolling window override for slots

supalarry reviewed (changes requested):
Thank you for your contribution!

I have a question though - why accept `disableRollingWindowAdjustment` as a property on the API level? I think that if rolling window is enabled and `startTime` is provided when slots are fetched we should set `disableRollingWindowAdjustment` to true ourselves.

What do you think? I think it should be like that because users simply want to get slots from X to Y and not have to think about what is the rolling window adjustement.

Good PR - looking to hear from you.

hackice20 committed:
Remove disableRollingWindowAdjustment from public API, auto-disable for API calls

hackice20 committed:
fix: remove unused rolling window adjustment logic from util.ts

Ryukemeister reviewed (changes requested):
hi @hackice20 , thank you for the contribution! can you also update the e2e tests at `apps/api/v2/src/modules/slots/slots-2024-09-04/controllers/e2e` to make sure we cover the rolling window adjustment use case. basically add a test that creates an event type with rolling window property set, and then we fetch slots for that event we need to make sure the start property is being respected.

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hackice20 committed:
test(slots): add e2e test for rolling window adjustment fix

- Add e2e test to verify rolling window adjustment is disabled for API calls
- Test ensures slots are returned only within requested date range
- Uses dynamic dates to avoid time bomb issues
- Add disableRollingWindowAdjustment to TRPC schema for type safety
- Restore conditional rolling window logic in util.ts (per bot feedback)
- Preserves backward compatibility: web/TRPC still get rolling window by default

supalarry commented on apps/api/v2/src/modules/slots/slots-2024-09-04/controllers/e2e/user-event-type-slots.controller.e2e-spec.ts:
```diff
@@ -1923,9 +1924,115 @@ describe("Slots 2024-09-04 Endpoints", () => {
+      });
+
+      it("should respect start property and return slots only within requested date range for rolling window event type", async () => {
+        // Use dynamic dates in the future to avoid "time bomb" issues
```
Can we remove this comment? The code implies the logic already

supalarry reviewed (changes requested):
Good job - we are getting there. Have last piece of feedback left. Let me know what you think.

hackice20 committed:
chore(slots): adjust rolling window docs and schema defaults

ThyMinimalDev committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

supalarry commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -1037,15 +1037,14 @@ export class AvailableSlotsService {
 
-    // If startTime is already sent in the past, we don't need to adjust it.
-    // We assume that the client is already sending startTime as per their requirement.
-    // Note: We could optimize it further to go back 1 month in past only for the 2nd month because that is what we are putting a hard limit at.
```
Why did we delete this comment??? Please leave as is and in this case there is no need for a new comment because constant disableRollingWindowAdjustment is self explanatory.

supalarry reviewed (changes requested):
after 2 comments below good to go

hackice20 committed:
fix: rolling window flag and comments

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hackice20 committed:
fix: wire rolling window flag in v2 and legacy slots

supalarry commented on apps/api/v2/src/modules/slots/slots-2024-09-04/controllers/e2e/user-event-type-slots.controller.e2e-spec.ts:
```diff
@@ -1923,9 +1924,93 @@ describe("Slots 2024-09-04 Endpoints", () => {
+        expect(typeof slots).toBe("object");
+
+        const returnedDates = Object.keys(slots);
+        expect(returnedDates.length).toBeGreaterThan(0);
```
Can we please change this to expect a specific amount of slots returned instead of greater than 0? It would make the expectations more defined.

supalarry reviewed (commented):
Great stuff man! Thank you. I have just 2 small comments about tests and then will write to @hariombalhara to check this out too.

hackice20 committed:
test: tighten rolling-window slots e2e expectations

supalarry commented:
> fixed both the things @supalarry

Amazing, thank you. Let's see what hariom says.

hackice20 commented:
> > fixed both the things @supalarry
> 
> Amazing, thank you. Let's see what hariom says.

🫡

hariombalhara commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -1045,7 +1046,7 @@ export class AvailableSlotsService {
     // Note: We could optimize it further to go back 1 month in past only for the 2nd month because that is what we are putting a hard limit at.
     const startTimeAdjustedForRollingWindowComputation =
-      isStartTimeInPast || !isRollingWindowPeriodType
+      isStartTimeInPast || !isRollingWindowPeriodType || disableRollingWindowAdjustment
```
Also, that handling of not returning slots before startDate could be okay for regular calls(non-api-v2) too, we might want to test that

hariombalhara reviewed (changes requested):
There is an issue with the implementation

hackice20 committed:
fix: keep rolling window anchored to today, add boundary test

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hackice20 committed:
fix: use disableRollingWindowAdjustment flag in slots computation

hackice20 committed:
Revert "fix: use disableRollingWindowAdjustment flag in slots computation"

This reverts commit 0ab59cded0dd71067368b51870cfd10ecc5fa391.

hackice20 committed:
fix: filter slots response by requested start date for API v2

hackice20 committed:
fix: remove unused disableRollingWindowAdjustment variable from util.ts

hackice20 commented:
> I don't see the variable in slots/util.ts being used. Infact I don't see anything else besides that in the file

I now have reverted `slots/util.ts` ,as it was unused.
in the `slots.service.ts` file the `filterSlotsByStartDate` method filters the slots response to only include dates >= the requested start date (e.g., if the request asks for `start=2025-12-09`, it removes any slots from `2025-11-26` that were included because rolling window computation looked back 1 month).
Please let me know how I can improve this. Thanking you in anticipation! @hariombalhara

hackice20 committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

ThyMinimalDev committed:
Merge branch 'main' into fix/v2-slots-rolling-window-date-range

hariombalhara commented:
Closing it.
